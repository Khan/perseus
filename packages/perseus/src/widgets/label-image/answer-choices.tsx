/**
 * Enables single or multiple answers selection using WonderBlocks dropdowns.
 */

import {
    MultiSelect,
    SingleSelect,
    OptionItem,
} from "@khanacademy/wonder-blocks-dropdown";
import * as React from "react";
import _ from "underscore";

import Renderer from "../../renderer";

export type AnswerType = {
    // The answer string, can be plain text or a KaTeX expression.
    content: string;
    // Whether the answer is selected.
    checked: boolean;
};

type AnswerChoicesProps = {
    // The list of possible answers in a specific order.
    choices: ReadonlyArray<AnswerType>;
    // Whether multiple answers may be chosen.
    multipleSelect: boolean;
    // Callback to handle change to answer choices selection.
    onChange: (selection: ReadonlyArray<boolean>) => void;
    // Callback to define custom opener.
    opener: SingleSelect["props"]["opener"];
    // Callback to handle toggle of dropdown.
    onToggle: (opened: boolean) => unknown;
    // Whether the answer choices are disabled.
    disabled: boolean;
};

const AnswerChoices = (props: AnswerChoicesProps) => {
    const onAnswerChange = (selected: string[]) => {
        const {choices, onChange} = props;
        onChange(choices.map((choice) => selected.includes(choice.content)));
    };

    // WB Dropdown types only take an array of nodes, so we can't use a
    // functional component, which can only return a single node.
    const AnswerItems = (choices: readonly AnswerType[]) =>
        choices.map(({content}) => (
            <OptionItem
                key={content}
                value={content}
                label={<Renderer content={content} inline />}
            />
        ));

    const selectedValues = props.choices
        .filter((choice) => choice.checked)
        .map((choice) => choice.content);

    const {opener, onToggle, disabled} = props;

    const args: Partial<
        React.ComponentProps<typeof MultiSelect> &
            React.ComponentProps<typeof SingleSelect>
    > = {
        // reset to allow child (answer pill) to control z-index
        style: {zIndex: "unset"},
        children: AnswerItems(props.choices),
        opener,
        onToggle,
        disabled,
    };

    return props.multipleSelect ? (
        <MultiSelect
            {...args}
            selectedValues={selectedValues}
            onChange={(selected) => onAnswerChange(selected)}
        />
    ) : (
        <SingleSelect
            {...args}
            selectedValue={selectedValues[0]}
            onChange={(selected) => onAnswerChange([selected])}
            placeholder={props.choices[0].content} // not visible, but required
        />
    );
};

export default AnswerChoices;
