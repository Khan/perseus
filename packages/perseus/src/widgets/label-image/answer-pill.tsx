import Color from "@khanacademy/wonder-blocks-color";
import {useUniqueIdWithMock} from "@khanacademy/wonder-blocks-core";
import * as i18n from "@khanacademy/wonder-blocks-i18n";
import Pill from "@khanacademy/wonder-blocks-pill";
import {StyleSheet, type CSSProperties} from "aphrodite";
import * as React from "react";
import {Popper} from "react-popper";

import Renderer from "../../renderer";

const BringToFront: CSSProperties = {
    boxShadow: `0 8px 8px ${Color.offBlack64}`,
    zIndex: 1000,
};

export const AnswerPill = (props: {
    selectedAnswers: readonly string[];
    showCorrectness?: "correct" | "incorrect";
    markerRef?: HTMLElement;
    side: "top" | "bottom" | "left" | "right";
    style?: CSSProperties;
    hovered?: boolean;
    focused?: boolean;
    onClick?: () => void;
}) => {
    const {
        selectedAnswers,
        showCorrectness,
        markerRef,
        side,
        onClick,
        style,
        focused,
        hovered,
    } = props;

    const idFactory = useUniqueIdWithMock();

    const answerString =
        selectedAnswers.length > 1
            ? // always need `ngettext` for variable numbers even if we don't use the singular, see https://khanacademy.slack.com/archives/C0918TZ5G/p1700163024293079
              i18n.ngettext(
                  "%(num)s answer",
                  "%(num)s answers",
                  selectedAnswers.length,
              )
            : selectedAnswers[0];

    // It should no longer be possible to interact with an answer after it
    // has been labeled (answered) correctly.
    const correct = showCorrectness === "correct";
    const incorrect = showCorrectness === "incorrect";

    return (
        <Popper
            placement={side}
            referenceElement={markerRef}
            modifiers={[
                {
                    name: "preventOverflow",
                    options: {
                        rootBoundary: "viewport",
                    },
                },
            ]}
        >
            {({ref, style: popperStyle}) => (
                <Pill
                    size="large"
                    kind="accent"
                    id={idFactory.get("perseus-label-image-widget-answer-pill")}
                    onClick={correct ? undefined : onClick}
                    ref={ref}
                    style={[
                        style,
                        popperStyle,
                        styles.pill,
                        correct && styles.correct,
                        incorrect && styles.incorrect,
                        (focused || hovered) && BringToFront,
                    ]}
                >
                    <Renderer content={answerString} inline />
                </Pill>
            )}
        </Popper>
    );
};

const styles = StyleSheet.create({
    correct: {
        // WB green darkened by 18%
        backgroundColor: "#00880b",
    },
    incorrect: {
        backgroundColor: Color.offBlack64,
    },
    pill: {
        // Reset the Pill's default height in order to account
        // for multi-line pills.
        height: "auto",
    },
});
