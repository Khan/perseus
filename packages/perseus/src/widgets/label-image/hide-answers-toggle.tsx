import {View, useUniqueIdWithMock} from "@khanacademy/wonder-blocks-core";
import Switch from "@khanacademy/wonder-blocks-switch";
import {LabelMedium} from "@khanacademy/wonder-blocks-typography";
import {StyleSheet} from "aphrodite";
import * as React from "react";

import {strings} from "./strings";

export const HideAnswersToggle = (props: {
    areAnswersHidden: boolean;
    onChange: React.ComponentPropsWithoutRef<typeof Switch>["onChange"];
}) => {
    const idFactory = useUniqueIdWithMock();
    const switchId = idFactory.get("hide-answers-toggle");
    const labelId = idFactory.get("hide-answers-label");
    return (
        <View style={styles.switchWrapper}>
            <Switch
                id={switchId}
                checked={props.areAnswersHidden}
                onChange={props.onChange}
                aria-labelledby={labelId}
            />
            <LabelMedium id={labelId} htmlFor={switchId} tag="label">
                {strings.hideAnswersToggleLabel}
            </LabelMedium>
        </View>
    );
};

const styles = StyleSheet.create({
    switchWrapper: {
        display: "flex",
        flexDirection: "row",
        flexWrap: "wrap-reverse",
        alignItems: "center",
        gap: "0.5em",
        marginTop: "1em",
    },
});
