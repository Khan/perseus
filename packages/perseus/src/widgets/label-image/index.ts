export {default} from "./label-image";
