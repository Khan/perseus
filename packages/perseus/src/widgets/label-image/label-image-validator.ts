import type {InteractiveMarkerType} from "./types";
import type {PerseusScore} from "../../types";
import type {
    PerseusLabelImageRubric,
    PerseusLabelImageUserInput,
} from "../../validation.types";

// Question state for marker as result of user selected answers.
type InteractiveMarkerScore = {
    // Whether user selected answers for the marker.
    hasAnswers: boolean;
    // Whether user (answer) selection answered the question correctly.
    isCorrect: boolean;
};

export function scoreMarker(
    marker: InteractiveMarkerType,
): InteractiveMarkerScore {
    const score = {
        hasAnswers: false,
        isCorrect: false,
    };

    if (marker.selected && marker.selected.length > 0) {
        score.hasAnswers = true;
    }

    if (marker.answers.length > 0) {
        if (
            marker.selected &&
            marker.selected.length === marker.answers.length
        ) {
            // All correct answers are selected by the user.
            score.isCorrect = marker.selected.every((choice) =>
                marker.answers.includes(choice),
            );
        }
    } else if (!marker.selected || marker.selected.length === 0) {
        // Correct as no answers should be selected by the user.
        score.isCorrect = true;
    }

    return score;
}

function labelImageValidator(
    state: PerseusLabelImageUserInput,
    rubric?: PerseusLabelImageRubric,
): PerseusScore {
    let numAnswered = 0;
    let numCorrect = 0;

    for (const marker of state.markers) {
        const score = scoreMarker(marker);

        if (score.hasAnswers) {
            numAnswered++;
        }

        if (score.isCorrect) {
            numCorrect++;
        }
    }

    // We expect all question markers to be answered before grading.
    if (numAnswered !== state.markers.length) {
        return {
            type: "invalid",
            message: null,
        };
    }

    return {
        type: "points",
        // Markers with no expected answers are graded as correct if user
        // makes no answer selection.
        earned: numCorrect === state.markers.length ? 1 : 0,
        total: 1,
        message: null,
    };
}

export default labelImageValidator;
