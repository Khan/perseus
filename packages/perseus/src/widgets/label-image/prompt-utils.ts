import type labelImage from "./label-image";
import type {WidgetType} from "../../prompt-types";
import type {PerseusLabelImageUserInput} from "../../validation.types";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type WidgetProps = PropsFor<typeof labelImage.widget>;

type BaseMarker = {
    label: WidgetProps["markers"][0]["label"];
};

type UserInputMarker = {
    label: PerseusLabelImageUserInput["markers"][0]["label"];
    selected: PerseusLabelImageUserInput["markers"][0]["selected"];
};

export type LabelImagePromptJSON = {
    type: WidgetType;
    options: {
        choices: WidgetProps["choices"];
        imageUrl: WidgetProps["imageUrl"];
        imageAlt: WidgetProps["imageAlt"];
        markers: BaseMarker[];
    };
    userInput: {
        markers: UserInputMarker[];
    };
};

export const getPromptJSON = (
    renderProps: WidgetProps,
    userInput: PerseusLabelImageUserInput,
): LabelImagePromptJSON => {
    const propMarkers = renderProps.markers.map((marker) => {
        return {
            label: marker.label,
        };
    });

    const inputMarkers = userInput.markers.map((marker) => {
        return {
            label: marker.label,
            selected: marker.selected,
        };
    });

    return {
        type: "label-image",
        options: {
            choices: renderProps.choices,
            imageUrl: renderProps.imageUrl,
            imageAlt: renderProps.imageAlt,
            markers: propMarkers,
        },
        userInput: {
            markers: inputMarkers,
        },
    };
};
