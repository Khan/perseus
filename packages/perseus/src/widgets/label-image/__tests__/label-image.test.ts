import {screen} from "@testing-library/react";
import {userEvent as userEventLib} from "@testing-library/user-event";

import {
    testDependencies,
    testDependenciesV2,
} from "../../../../../../testing/test-dependencies";
import * as Dependencies from "../../../dependencies";
import {scorePerseusItemTesting} from "../../../util/test-utils";
import {renderQuestion} from "../../__testutils__/renderQuestion";
import {LabelImage} from "../label-image";

import {
    shortTextQuestion,
    textQuestion,
    textWithoutAnswersQuestion,
} from "./label-image.testdata";

import type {UserEvent} from "@testing-library/user-event";

const emptyMarker = {
    label: "",
    answers: [],
    selected: [],
    x: 0,
    y: 0,
} as const;

describe("LabelImage", function () {
    let userEvent: UserEvent;
    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });

        // Mocked for loading graphie in svg-image
        global.fetch = jest.fn(() =>
            Promise.resolve({
                text: () => "",
                ok: true,
            }),
        ) as jest.Mock;
    });

    describe("imageSideForMarkerPosition", function () {
        it("should return side: top", function () {
            expect(
                LabelImage.imageSideForMarkerPosition(21, 0, "NONE"),
            ).toEqual("top");
            expect(
                LabelImage.imageSideForMarkerPosition(79, 0, "NONE"),
            ).toEqual("top");
            expect(
                LabelImage.imageSideForMarkerPosition(50, 0, "NONE"),
            ).toEqual("top");
            expect(
                LabelImage.imageSideForMarkerPosition(50, 25, "NONE"),
            ).toEqual("top");
            expect(
                LabelImage.imageSideForMarkerPosition(50, 50, "NONE"),
            ).toEqual("top");
        });

        it("should return side: right", function () {
            expect(
                LabelImage.imageSideForMarkerPosition(100, 0, "NONE"),
            ).toEqual("right");
            expect(
                LabelImage.imageSideForMarkerPosition(100, 100, "NONE"),
            ).toEqual("right");
            expect(
                LabelImage.imageSideForMarkerPosition(100, 50, "NONE"),
            ).toEqual("right");
            expect(
                LabelImage.imageSideForMarkerPosition(75, 50, "NONE"),
            ).toEqual("right");
            expect(
                LabelImage.imageSideForMarkerPosition(51, 50, "NONE"),
            ).toEqual("right");
        });

        it("should return side: bottom", function () {
            expect(
                LabelImage.imageSideForMarkerPosition(21, 100, "NONE"),
            ).toEqual("bottom");
            expect(
                LabelImage.imageSideForMarkerPosition(79, 100, "NONE"),
            ).toEqual("bottom");
            expect(
                LabelImage.imageSideForMarkerPosition(50, 100, "NONE"),
            ).toEqual("bottom");
            expect(
                LabelImage.imageSideForMarkerPosition(50, 75, "NONE"),
            ).toEqual("bottom");
            expect(
                LabelImage.imageSideForMarkerPosition(50, 51, "NONE"),
            ).toEqual("bottom");
        });

        it("should return side: left", function () {
            expect(LabelImage.imageSideForMarkerPosition(0, 0, "NONE")).toEqual(
                "left",
            );
            expect(
                LabelImage.imageSideForMarkerPosition(0, 100, "NONE"),
            ).toEqual("left");
            expect(
                LabelImage.imageSideForMarkerPosition(0, 50, "NONE"),
            ).toEqual("left");
            expect(
                LabelImage.imageSideForMarkerPosition(25, 50, "NONE"),
            ).toEqual("left");
            expect(
                LabelImage.imageSideForMarkerPosition(49, 50, "NONE"),
            ).toEqual("left");
        });
    });

    describe("navigateToMarkerIndex", function () {
        it("should not navigate if all other markers were answered", function () {
            const markers = [
                {
                    ...emptyMarker,
                    label: "this",
                },
                {
                    ...emptyMarker,
                    label: "above",
                    y: -1,
                    showCorrectness: "correct",
                },
                {
                    ...emptyMarker,
                    label: "right",
                    x: 1,
                    showCorrectness: "correct",
                },
                {
                    ...emptyMarker,
                    label: "below",
                    y: 1,
                    showCorrectness: "correct",
                },
                {
                    ...emptyMarker,
                    label: "left",
                    x: -1,
                    showCorrectness: "correct",
                },
            ];

            expect(
                markers[
                    // @ts-expect-error - TS2345 - Argument of type '({ label: string; answers: readonly []; selected: readonly []; x: 0; y: 0; } | { label: string; y: number; showCorrectness: string; answers: readonly []; selected: readonly []; x: 0; } | { label: string; x: number; showCorrectness: string; answers: readonly []; selected: readonly []; y: 0; })[]' is not assignable to parameter of type 'readonly InteractiveMarkerType[]'.
                    LabelImage.navigateToMarkerIndex({x: 0, y: 0}, markers, 0)
                ].label,
            ).toEqual("this");
        });

        describe("up direction", function () {
            const navigateDirecton = {x: 0, y: -1} as const;

            it("should navigate to marker above", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "below",
                        y: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "above",
                        y: -1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("above");
            });

            it("should navigate to lowest marker", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "lowest",
                        y: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "below",
                        y: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("lowest");
            });

            it("should navigate to the next above marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "above 1",
                        y: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "above 2",
                        y: -1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("above 1");
            });

            it("should navigate to the next coplanar marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 1",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 2",
                        y: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("coplanar 1");
            });
        });

        describe("right direction", function () {
            const navigateDirecton = {x: 1, y: 0} as const;

            it("should navigate to marker on the right", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "left",
                        x: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "right",
                        x: 1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("right");
            });

            it("should navigate to leftmost marker", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "leftmost",
                        x: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "left",
                        x: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("leftmost");
            });

            it("should navigate to the next right marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "right 1",
                        x: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "right 2",
                        x: 1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("right 1");
            });

            it("should navigate to the next coplanar marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 1",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 2",
                        x: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("coplanar 1");
            });
        });

        describe("down direction", function () {
            const navigateDirecton = {x: 0, y: 1} as const;

            it("should navigate to marker below", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "above",
                        y: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "below",
                        y: 1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("below");
            });

            it("should navigate to highest marker", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "highest",
                        y: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "above",
                        y: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("highest");
            });

            it("should navigate to the next below marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "below 1",
                        y: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "below 2",
                        y: 1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("below 1");
            });

            it("should navigate to the next coplanar marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 1",
                        y: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 2",
                        y: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("coplanar 1");
            });
        });

        describe("left direction", function () {
            const navigateDirecton = {x: -1, y: 0} as const;

            it("should navigate to marker on the left", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "right",
                        x: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "left",
                        x: -1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("left");
            });

            it("should navigate to rightmost marker", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "rightmost",
                        x: 1,
                    },
                    {
                        ...emptyMarker,
                        label: "right",
                        x: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("rightmost");
            });

            it("should navigate to the next left marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "left 1",
                        x: -1,
                    },
                    {
                        ...emptyMarker,
                        label: "left 2",
                        x: -1,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("left 1");
            });

            it("should navigate to the next coplanar marker in original order", function () {
                const markers = [
                    {
                        ...emptyMarker,
                        label: "this",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 1",
                        x: 0,
                    },
                    {
                        ...emptyMarker,
                        label: "coplanar 2",
                        x: 0,
                    },
                ];

                expect(
                    markers[
                        LabelImage.navigateToMarkerIndex(
                            navigateDirecton,
                            markers,
                            0,
                        )
                    ].label,
                ).toEqual("coplanar 1");
            });
        });
    });

    describe("analytics", () => {
        beforeEach(() => {
            jest.spyOn(testDependenciesV2.analytics, "onAnalyticsEvent");
            jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
                testDependencies,
            );
            jest.spyOn(Dependencies, "useDependencies").mockReturnValue(
                testDependenciesV2,
            );
        });

        it("sends an analytics event when the toggle is interacted with", async () => {
            // render component
            renderQuestion(textQuestion);

            // Toggle the switch
            const toggleAnswerSwitch = await screen.findByRole("switch");
            await userEvent.click(toggleAnswerSwitch);

            expect(
                testDependenciesV2.analytics.onAnalyticsEvent,
            ).toHaveBeenCalledWith({
                type: "perseus:label-image:toggle-answers-hidden",
                payload: null,
            });
        });

        it("sends an analytics event when a marker is interacted with", async () => {
            // render component
            renderQuestion(textQuestion);

            // Toggle the button
            const markerButton = await screen.findByLabelText(
                "The fourth unlabeled bar line.",
            );
            await userEvent.click(markerButton);

            expect(
                testDependenciesV2.analytics.onAnalyticsEvent,
            ).toHaveBeenCalledWith({
                type: "perseus:label-image:marker-interacted-with",
                payload: null,
            });
        });

        it("sends analytics when a choice is interacted with", async () => {
            // render component
            renderQuestion(textQuestion);

            // Toggle the marker
            const markerButton = await screen.findByLabelText(
                "The fourth unlabeled bar line.",
            );
            await userEvent.click(markerButton);

            // Select a choice
            const choices = await screen.findAllByText("Trucks");
            const choice = choices[choices.length - 1];
            await userEvent.click(choice);

            expect(
                testDependenciesV2.analytics.onAnalyticsEvent,
            ).toHaveBeenCalledWith({
                type: "perseus:label-image:choiced-interacted-with",
                payload: null,
            });
        });
    });

    describe("getUserInput", () => {
        it("should return the current user input on initial render", () => {
            // render component
            const {renderer} = renderQuestion(textQuestion);

            const userInput = renderer.getUserInputMap();

            expect(userInput).toEqual({
                "label-image 1": {
                    markers: [
                        {
                            label: "The fourth unlabeled bar line.",
                            selected: undefined,
                        },
                        {
                            label: "The third unlabeled bar line.",
                            selected: undefined,
                        },
                        {
                            label: "The second unlabeled bar line.",
                            selected: undefined,
                        },
                        {
                            label: "The first unlabeled bar line.",
                            selected: undefined,
                        },
                    ],
                },
            });
        });
    });

    describe("scorePerseusItem", () => {
        it("should be invalid on first render", () => {
            // Arrange
            const {renderer} = renderQuestion(textQuestion);

            // Act
            const score = scorePerseusItemTesting(
                textQuestion,
                renderer.getUserInputMap(),
            );

            // Assert
            expect(score).toHaveInvalidInput();
        });

        it("can be answered correctly when correct option is picked for the marker", async () => {
            // Arrange
            const {renderer} = renderQuestion(shortTextQuestion);

            // Act
            const markerButton = screen.getByRole("button", {
                name: "The fourth unlabeled bar line.",
            });
            await userEvent.click(markerButton);

            const choice = screen.getByRole("option", {name: "SUVs"});
            await userEvent.click(choice);

            const score = scorePerseusItemTesting(
                shortTextQuestion,
                renderer.getUserInputMap(),
            );

            // Assert
            expect(score).toHaveBeenAnsweredCorrectly();
        });

        it("can be answered incorrectly when incorrect option picked for the marker", async () => {
            // Arrange
            const {renderer} = renderQuestion(shortTextQuestion);

            // Act
            const markerButton = screen.getByRole("button", {
                name: "The fourth unlabeled bar line.",
            });
            await userEvent.click(markerButton);

            const choice = screen.getByRole("option", {name: "Trucks"});
            await userEvent.click(choice);

            const score = scorePerseusItemTesting(
                shortTextQuestion,
                renderer.getUserInputMap(),
            );

            // Assert
            expect(score).toHaveBeenAnsweredIncorrectly();
        });
    });

    describe("textWithoutAnswersQuestion", () => {
        it("should render the widget without answers", async () => {
            // Arrange
            renderQuestion(textWithoutAnswersQuestion);

            // Act and Assert
            const markerButton = screen.getByRole("button", {
                name: "The fourth unlabeled bar line.",
            });
            // Confirms the widget renders and that marker buttons are present
            await userEvent.click(markerButton);
        });
    });
});
