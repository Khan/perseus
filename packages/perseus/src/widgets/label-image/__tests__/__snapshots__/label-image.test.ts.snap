// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textWithoutAnswersQuestion should render a text question without answers 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Carol created a chart and a bar graph to show how many of each type of vehicle were in her supermarket parking lot.
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <table>
        <thead>
          <tr>
            <th
              scope="col"
              style="text-align: left;"
            >
              Vehicle Type
            </th>
            <th
              scope="col"
              style="text-align: center;"
            >
              Number in the parking lot
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              style="text-align: left;"
            >
              Trucks
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  25
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: left;"
            >
              Vans
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  5
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: left;"
            >
              Cars
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  40
                </span>
                <span />
              </span>
            </td>
          </tr>
          <tr>
            <td
              style="text-align: left;"
            >
              SUVs
            </td>
            <td
              style="text-align: center;"
            >
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  10
                </span>
                <span />
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        <strong>
          Label each bar on the bar graph.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="3"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <svg
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
          >
            <title>
              Rendering Error!
            </title>
            <path
              d="m8 16c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm0-3c0.55 0 1-0.45 1-1s-0.45-1-1-1-1 0.45-1 1 0.45 1 1 1zm0-9c-0.55 0-1 0.45-1 1v4c0 0.55.45 1 1 1s1-0.45 1-1v-4c0-0.55-0.45-1-1-1z"
              fill="#d92916"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
